### PR TITLE
Use annotation to mark DecoupleIgnore

### PR DESCRIPTION
**Pull request by kgyrtkirk**
Its a bit time consuming to check newly passing tests/etc; in #14996 the question came up "since when" some of those tests are passing....

This changeset introduces an annotation to mark disabled tests in decoupled mode - which has a few benefits:

the main difference is that they will still be executed ; but the expected error is classified.
In case it fails with the expected issue - it will put an ignore on the test result; however if the behaviour changes - it will fail the test.

### CODE REVIEW

File: sql/src/test/java/org/apache/druid/sql/calcite/DecoupledIgnore.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.regex.Pattern;
+
+/**
+ * Can be used to mark tests which are not-yet supported in decoupled mode.
+ */
```
Comment (imply-cheddar) [inline sql/src/test/java/org/apache/druid/sql/calcite/DecoupledIgnore.java:32]: Perhaps add extra words here that describe what someone should do if they suddenly see a `DecoupledIgnore` test that now starts to succeed.

**Review by imply-cheddar — APPROVED** [review]
I love this, proper negative tests!

File: sql/src/test/java/org/apache/druid/sql/calcite/DecoupledIgnore.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    Modes(Class<? extends Throwable> cl, String regex)
+    {
+      this.throwableClass = cl;
+      this.regex = regex;
+    }
+
+    Pattern getPattern()
+    {
+      return Pattern.compile(regex);
```
Comment (rohangarg) [inline sql/src/test/java/org/apache/druid/sql/calcite/DecoupledIgnore.java:68]: small suggestion : if the `compile` take considerable cumulative time in the test run, we could move this to constructor too.

**Review by rohangarg — APPROVED** [review]
Thanks so much for the changes @kgyrtkirk - LGTM!

File: sql/src/test/java/org/apache/druid/sql/calcite/DecoupledIgnore.java (hunk 2 above)
Comment (kgyrtkirk) [reply sql/src/test/java/org/apache/druid/sql/calcite/DecoupledIgnore.java:68]: Actually that was my intention when I've introduced this method - but I've forgot it.

The current patterns doesn't seem to be that costly; can we make this change later - when we add more patterns/etc; its hard to get clear testruns nowadays...

Comment (rohangarg) [reply sql/src/test/java/org/apache/druid/sql/calcite/DecoupledIgnore.java:68]: Yes, sure that makes sense

**Pull request merged**